### DISCUSSION TITLE
Which E2SM KPM version does srsRAN use?

### DISCUSSION THREAD

**Discussion by P4sca1** [srsran/srsRAN_Project#596] [General Help] [answered]
I compiled FlexRIC using the `-DKPM_VERSION=KPM_V3` flag.  However, FlexRIC cashes when it tries to decode a RICIndication message from srsRAN gNodeB:
```
nearRT-RIC: /flexric/src/sm/kpm_sm/kpm_sm_v03.00/dec/kpm_dec_asn.c:186: kpm_dec_ind_msg_asn: Assertion `rval.code == RC_OK && "Are you sending data in ATS_ALIGNED_BASIC_PER syntax?"' failed.
```

I checked the gnb_e2ap.pcap file which indicates that KPM version 2 is used, instead of version 3.
<img width="1904" alt="Bildschirmfoto 2024-05-02 um 14 27 55" src="https://github.com/srsran/srsRAN_Project/assets/10992664/2a3f296b-3bcc-4bcc-a5f1-989982a62e42">

Based on the following statement, I expected srsRAN to use version 3.
> In addition, the repository provides two example Python-based xApps that allow requesting and monitoring E2 metrics from the gNB using all reporting styles (1-5) of the e2sm-kpm module (which was updated to version 3).
https://github.com/srsran/srsRAN_Project/issues/471#issuecomment-2049826630

Also, am I correct that the version refers to the O-RAN specification of E2SM KPM? Which means that version 4 is already standardized? 
<img width="1302" alt="Bildschirmfoto 2024-05-02 um 14 29 41" src="https://github.com/srsran/srsRAN_Project/assets/10992664/1831a10a-86a5-4e47-b023-907db54ddee5">

**Comment by pgawlowicz** [comment, marked as answer]
Please check this patch: https://github.com/srsran/srsRAN_Project/discussions/567#discussioncomment-9209950

**Reply by P4sca1** [reply]
I tried with the patch file and it is working now. It is important to start the xApp after the UE connects. Otherwise you get the following error:
```
nearRT-RIC: /flexric/src/sm/kpm_sm/kpm_sm_v03.00/dec/dec_asn_kpm_common/dec_meas_data.c:20: kpm_dec_meas_data_asn: Assertion `meas_data[i].meas_record_len >= 1 && meas_data[i].meas_record_len <= maxnoofMeasurementValue' failed.
```
Could you explain why Wireshark still shows KPM v2 in the RICIndication messages instead of v3?

**Reply by pgawlowicz** [reply]
Thanks for pointing this out, I will check. Probably, we send an empty RIC indication, while we should set the reported values to 0. 

Not sure, about Wireshark. Which version do you use? You might try to compile from source.

**Reply by P4sca1** [reply]
I use version 4.2.4 on MacOS. Just wanted to make sure, that srsRAN sends version 3 packets.

**Reply by pgawlowicz** [reply]
Not sure why Wireshark shows `KPM v2`, it seems the dev version uses already v3 (check [here](https://github.com/wireshark/wireshark/blob/master/epan/dissectors/asn1/e2ap/e2sm-kpm-v3.00.asn#L7)).  srsRAN also uses version 3. 

Note that the in version 2, the Timestamp has 4 bytes (see [here](https://github.com/wireshark/wireshark/blob/b4bcf691661a975a90ba736040f44f2a6f0b486e/epan/dissectors/asn1/kpm-v2/e2sm-kpm-v2.02.asn#L31)), 
while in version 3, it has 8 bytes (see [here](https://github.com/wireshark/wireshark/blob/master/epan/dissectors/asn1/e2ap/e2sm-kpm-v3.00.asn#L31))

**Comment by yagoda** [comment]
Hi @P4sca1,

In preparation for our upcoming workshop in DC, we are contacting some users of our E2 solution. Can you please share a few lines about your experience with the E2 interface, what you were trying to achieve, and how it worked out for you?

Kind Regards,
Justin

**Reply by P4sca1** [reply]
Hey @yagoda,

I am currently on vacation, so I can't give a detailed answer.
I was using srsRAN to build a simulated 5G network using Open5GS, FlexRIC / O-RAN SC RIC, srsRAN gNB, zeroMQ, and srsUE as part of my bachelor thesis. 
A monitoring xApp was deployed to capture UE metrics.
You can find the thesis here: https://www.dn.th-koeln.de/wp-content/uploads/2024/06/Bachelorarbeit-Pascal_Sthamer-final.pdf
